crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs

```diff
@@ -0,0 +1,469 @@
+    pub authority_state: Arc<AuthorityState>,
+    pub package: ObjectRef,
+    pub upgrade_cap: ObjectRef,
+    pub framework_object_ref: ObjectRef,
+}
+
+impl UpgradeStateRunner {
+    pub async fn new(base_package_name: &str) -> Self {
+        let _dont_remove = ProtocolConfig::apply_overrides_for_testing(|_, mut config| {
+            config.set_package_upgrades_for_testing(true);
+            config
+        });
+        let (sender, sender_key): (_, AccountKeyPair) = get_key_pair();
+        let gas_object_id = ObjectID::random();
+        let gas_object = Object::with_id_owner_gas_for_testing(gas_object_id, sender, 100000);
+        let authority_state = init_state().await;
+        authority_state.insert_genesis_object(gas_object).await;
+
+        let (package, upgrade_cap) = build_and_publish_test_package_with_upgrade_cap(
+            &authority_state,
+            &sender,
+            &sender_key,
+            &gas_object_id,
+            base_package_name,
+            true,
+        )
+        .await;
+
+        let framework_object_ref = authority_state.get_framework_object_ref().await.unwrap();
+        Self {
+            sender,
+            sender_key,
+            gas_object_id,
+            authority_state,
+            package,
+            upgrade_cap,
+            framework_object_ref,
+        }
+    }
+
+    pub async fn run(&self, pt: ProgrammableTransaction) -> TransactionEffects {
+        execute_programmable_transaction(
+            &self.authority_state,
+            &self.gas_object_id,
+            &self.sender,
+            &self.sender_key,
+            pt,
+        )
+        .await
+        .unwrap()
+    }
+}
+
+#[tokio::test]
+async fn test_upgrade_package_happy_path() {
+    let runner = UpgradeStateRunner::new("move_upgrade/base").await;
+    let pt = {
+        let mut builder = ProgrammableTransactionBuilder::new();
+        let current_package_id = runner.package.0;
+        let (digest, modules) = build_upgrade_test_modules("stage1_basic_compatibility_valid");
+
+        // We take as input the upgrade cap
+        builder
+            .obj(ObjectArg::ImmOrOwnedObject(runner.upgrade_cap))
+            .unwrap();
+
+        // Create the upgrade ticket
+        let upgrade_arg = builder.pure(UPGRADE_POLICY_COMPATIBLE).unwrap();
+        let digest_arg = builder.pure(digest).unwrap();
+        let upgrade_ticket = move_call! {
+            builder,
+            (runner.framework_object_ref.0)::package::authorize_upgrade(Argument::Input(0), upgrade_arg, digest_arg)
+        };
+        let upgrade_receipt = builder.upgrade(current_package_id, upgrade_ticket, vec![], modules);
+        move_call! {
+            builder,
+            (runner.framework_object_ref.0)::package::commit_upgrade(Argument::Input(0), upgrade_receipt)
+        };
+
+        builder.finish()
+    };
+    let TransactionEffects::V1(effects) = runner.run(pt).await;
+
+    assert!(effects.status.is_ok());
```

Comment:
We should also fetch the upgraded package out from the store and check that it's kind of what we expected (maybe check its normalized modules for the functions that you changed?)

Reply:
Added :)